nucypher/blockchain/eth/token.py

```diff
@@ -124,43 +127,89 @@ def __str__(self) -> str:
 
 class Stake:
     """
-    A quantity of tokens, and staking time-frame for one stake for one miner.
+    A quantity of tokens and staking duration for one stake for one miner.
     """
 
+    class StakingError(Exception):
+        """Raised when a staking operation cannot be executed due to failure."""
+
     __ID_LENGTH = 16
 
     def __init__(self,
-                 owner_address: str,
-                 index: int,
+                 miner,
                  value: NU,
                  start_period: int,
-                 end_period: int):
+                 end_period: int,
+                 index: int,
+                 validate_now: bool = True):
 
-        # Stake Info
+        self.miner = miner
+        owner_address = miner.checksum_public_address
+        self.log = Logger(f'stake-{owner_address}-{index}')
+
+        # Stake Metadata
         self.owner_address = owner_address
         self.index = index
         self.value = value
         self.start_period = start_period
         self.end_period = end_period
-        self.duration = (self.end_period-self.start_period) + 1
 
-        # Internals
+        # Time
         self.start_datetime = datetime_at_period(period=start_period)
         self.end_datetime = datetime_at_period(period=end_period)
         self.duration_delta = self.end_datetime - self.start_datetime
 
-    def __repr__(self):
+        self.blockchain = miner.blockchain
+
+        # Agency
+        self.miner_agent = miner.miner_agent
+        self.token_agent = miner.token_agent
+
+        # Economics
+        self.economics = miner.economics
+        self.minimum_nu = NU(int(self.economics.minimum_allowed_locked), 'NuNit')
+        self.maximum_nu = NU(int(self.economics.maximum_allowed_locked), 'NuNit')
+
+        if validate_now:
+            self.validate_duration()
+
+        self.transactions = NO_STAKING_RECEIPT
+        self.receipt = NO_STAKING_RECEIPT
+
+    def __repr__(self) -> str:
         r = f'Stake(index={self.index}, value={self.value}, end_period={self.end_period})'
         return r
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
         return bool(self.value == other.value)
 
+    #
+    # Metadata
+    #
+
+    @property
+    def is_expired(self) -> bool:
+        current_period = self.miner_agent.get_current_period()
+        if current_period >= self.end_period:
```

Comment:
Is this more readable than `return current_period >= self.end_period`?

Reply:
:heavy_check_mark: